### PR TITLE
Add Automatic Gitignore Support

### PR DESCRIPTION
**Pull request by technicalpickles**
## Summary

Adds automatic `.gitignore` support to `pks`, respecting gitignored files during analysis. **Enabled by default** with option to disable for packwerk-identical behavior.

## Motivation

Previously, `pks` analyzed all files regardless of `.gitignore` status, including build artifacts (`node_modules/`), temporary files (`scratch/`), and logs. This meant:
- Unnecessary analysis of ignored files
- Manual exclusion patterns needed
- Behavior inconsistent with git-aware tools

## What Changed

**For users with `.gitignore` files:**
- Gitignored files are now automatically skipped
- Performance improvement from pruning ignored directories early
- No manual exclusion configuration needed

**Configuration:**
```yaml
# packwerk.yml - disable to match packwerk behavior
respect_gitignore: false  # default: true
```

**Supports:**
- Local `.gitignore` files
- Global gitignore (`~/.gitignore_global` or `core.excludesFile`)
- `.git/info/exclude`
- All standard patterns (wildcards, negation, etc.)

## Implementation

- Uses `ignore` crate (same as ripgrep) for battle-tested performance
- Prunes ignored directories during traversal (no stat calls)
- Gracefully handles missing/malformed gitignore files
- Works correctly in non-git directories

## Testing

- **10 integration tests** covering core scenarios
- **7 unit tests** for helper functions  
- Test fixtures validate common patterns and edge cases
- All existing tests pass (no regressions)

## Troubleshooting

**Files unexpectedly ignored?**
```bash
# Check what's matching
git check-ignore -v path/to/file.rb

# Temporarily disable
echo "respect_gitignore: false" >> packwerk.yml
```

See [ADVANCED_USAGE.md](ADVANCED_USAGE.md) for detailed documentation.

## Backward Compatibility

✅ Existing projects work without changes  
✅ Can be disabled if needed  
✅ Non-git directories continue working  
⚠️ Behavior change: gitignored files now skipped (generally desirable)

---

**Ready for review!** This brings `pks` behavior in line with git-aware tooling while providing an escape hatch for compatibility.

### CODE REVIEW

**Comment by perryqh** [comment]
Performance looks good!

Screenshot 2025-10-02 at 16 22 27

File: src/packs/walk_directory.rs [hunk 1]
```diff
@@ -27,6 +28,123 @@ impl jwalk::ClientState for ProcessReadDirState {
...
+    PathBuf::from(path)
+}
+
+/// Attempts to locate the global gitignore file.
+///
+/// This function checks multiple locations in order:
+/// 1. Git config `core.excludesFile` setting
+/// 2. `~/.gitignore_global`
+/// 3. `~/.config/git/ignore`
+///
```
Comment (perryqh) [inline src/packs/walk_directory.rs:54]: Not a blocker, but `excludesfile` can be set to anything in `~/.gitconfig`

File: src/packs/walk_directory.rs [hunk 2]
```diff
@@ -27,6 +28,123 @@ impl jwalk::ClientState for ProcessReadDirState {
...
+    // Try to read from git config first
+    if let Ok(output) = std::process::Command::new("git")
+        .args(["config", "--global", "core.excludesFile"])
+        .output()
+    {
+        if output.status.success() {
+            let path_str =
+                String::from_utf8_lossy(&output.stdout).trim().to_string();
+            if !path_str.is_empty() {
+                let expanded = expand_tilde(&path_str);
```
Comment (perryqh) [inline src/packs/walk_directory.rs]: Ah! The whole purpose of `expand_tilde` is to support `core.excludesFile`

File: src/packs/walk_directory.rs [hunk 3]
```diff
@@ -116,6 +251,19 @@ pub(crate) fn walk_directory(
                         let relative_path = child_absolute_dirname
                             .strip_prefix(cloned_absolute_root.as_ref())
                             .unwrap();
+
+                        // Check gitignore first (if enabled)
+                        if let Some(ref gitignore) = cloned_gitignore.as_ref() {
+                            let is_dir = child_dir_entry.file_type.is_dir();
```
Comment (perryqh) [inline src/packs/walk_directory.rs:210]: Given the check below:
```ruby
  if gitignore.matched(&relative_path, false).is_ignore() {
      continue;
  }
```
should we even attempt check is_ignore on line 260 if is_dir == false?

File: src/packs/walk_directory.rs (hunk 2 above)
Comment (technicalpickles) [reply src/packs/walk_directory.rs]: Yes! Added comments explaining that to the function

File: src/packs/walk_directory.rs (hunk 3 above)
Comment (technicalpickles) [reply src/packs/walk_directory.rs:210]: Good catch! Implemented in 8387c3b. The first gitignore check now only runs when is_dir == true since:
- Setting read_children_path = None on files is a no-op (files have no children)
- Files are already checked in the main loop (line ~304)
- Eliminates redundant matched() calls

**Review by dduugg — APPROVED** [review]
## Review

Overall this is well-scoped and well-tested. A few observations:

### Concerns

**1. Breaking behavior change by default** — The default of `true` means existing users silently get different results without any opt-in. Projects that previously analyzed files in gitignored paths (e.g. vendored code, generated files that someone actually wants checked) will get different output without knowing why. This is the right long-term default, but worth calling out clearly in release notes, and potentially adding a one-time warning on first run.

**2. Subprocess on every run** — `get_global_gitignore()` spawns `git config --global core.excludesFile` via `std::process::Command` on every `walk_directory` call. Gracefully returns `None` when git isn't on PATH, which is correct. Worth noting that the `ignore` crate already knows how to locate the global gitignore internally — see concern #4.

**3. Double `Arc` wrapping** — The code wraps `Option<Arc<Gitignore>>` in another `Arc`, resulting in `Arc<Option<Arc<Gitignore>>>`. The inner `Arc` is unnecessary; `Arc<Option<Gitignore>>` suffices.

**4. Missed opportunity: `ignore::WalkBuilder`** — The `ignore` crate has a high-level `WalkBuilder` that handles gitignore-aware traversal natively (it's what ripgrep uses). The PR manually integrates the lower-level `Gitignore` type with `jwalk`, which works fine. If the project ever reconsiders its parallelism strategy, `WalkBuilder` would handle all of this automatically.

**5. Pinned dependencies** — `globset = "=0.4.16"` and `ignore = "=0.4.23"` are pinned to avoid a Rust edition 2024 requirement in newer patches. Reasonable workaround, but these won't receive patch updates via `cargo update`. Worth a follow-up to upgrade to Rust edition 2024 and unpin.

### Minor Notes

- **`walk_directory` visibility change** (`pub(crate)` → `pub`): Done to allow `gitignore_test.rs` to import `build_gitignore_matcher`. Fine, but slightly expands the public API surface.
- **`per_file_cache.rs` fix**: The `create_dir_all` fix is correct and legitimate — just worth noting it's a bundled unrelated bug fix.
- **Test fixtures**: Well-structured. The `app_with_gitignore_disabled` fixture correctly validates opt-out behavior.

None of these are blockers. The feature is solid, the `ignore` crate is battle-tested, and the test coverage is thorough. Main ask is to document the behavioral change prominently in release notes.

**Comment by technicalpickles** [comment]
Thanks for the thorough review @dduugg\! All five concerns addressed:

**1. Breaking change documentation** — Added `CHANGELOG.md` documenting the behavioral change, who's affected, and the opt-out (`respect_gitignore: false` in `packwerk.yml`). Skipped the one-time runtime warning — the changelog entry covers it without adding runtime noise.

**2+4. Subprocess and WalkBuilder** — `get_global_gitignore()` no longer spawns a subprocess. It now calls `ignore::gitignore::gitconfig_excludes_path()` directly, which reads `~/.gitconfig` (and `$XDG_CONFIG_HOME/git/config`) without invoking `git`.

Regarding `WalkBuilder`: evaluated it, but it can't replace `jwalk` here. pks uses a custom `ProcessReadDirState` with jwalk's `ClientState` trait to propagate `current_package_yml` from parent to child directories in a single O(1) pass per entry. WalkBuilder's visitor pattern has no equivalent hook for carrying per-directory state across levels. Added a code comment in `walk_directory.rs` explaining the tradeoff so the next person doesn't have to re-derive it.

**3. Double Arc wrapping** — Fixed. `build_gitignore_matcher()` now returns `Option<Gitignore>` (no inner Arc), and the call site wraps it once in `Arc::new()` where needed.

**5. Pinned dependencies** — Unpinned. `rust-toolchain.toml` is already at 1.92.0 which supports edition 2024, so `ignore` and `globset` are now on semver ranges (`"0.4"`) and will pick up patch updates normally.

---

Smoke-tested against zenpayroll (46k files): zero behavioral delta with `respect_gitignore: true` vs `false` on the real codebase. The packwerk.yml exclude block already covers the same directories gitignore would filter. Feature verified correct with a synthetic `coverage/fake_service.rb` test file.
